tests/components/songpal/test_media_player.py

```diff
@@ -0,0 +1,237 @@
+"""Test songpal media_player."""
+import pytest
+from songpal import (
+    ConnectChange,
+    ContentChange,
+    PowerChange,
+    SongpalException,
+    VolumeChange,
+)
+
+from homeassistant.components import songpal
+from homeassistant.components.songpal.const import SET_SOUND_SETTING
+from homeassistant.components.songpal.media_player import (
+    STATE_OFF,
+    STATE_ON,
+    SUPPORT_SONGPAL,
+    SongpalDevice,
+    async_setup_entry,
+)
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers import device_registry as dr
+from homeassistant.setup import async_setup_component
+
+from . import (
+    CONF_DATA,
+    CONF_ENDPOINT,
+    CONF_NAME,
+    ENDPOINT,
+    FRIENDLY_NAME,
+    MAC,
+    MODEL,
+    SW_VERSION,
+    _create_mocked_device,
+    _patch_media_player_device,
+)
+
+from tests.async_mock import AsyncMock, MagicMock, call, patch
+from tests.common import MockConfigEntry
+
+
+async def test_setup_platform(hass):
+    """Test the legacy setup platform."""
+    mocked_device = _create_mocked_device(throw_exception=True)
+    with _patch_media_player_device(mocked_device):
+        await async_setup_component(
+            hass,
+            "media_player",
+            {
+                "media_player": [
+                    {
+                        "platform": songpal.DOMAIN,
+                        CONF_NAME: FRIENDLY_NAME,
+                        CONF_ENDPOINT: ENDPOINT,
+                    }
+                ],
+            },
+        )
+        await hass.async_block_till_done()
+
+    # No device is set up
+    mocked_device.assert_not_called()
+
+
+async def test_setup_failed(hass):
+    """Test failed to set up the entity."""
+    mocked_device = _create_mocked_device(throw_exception=True)
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    mocked_add_entity = MagicMock()
+
+    with _patch_media_player_device(mocked_device), pytest.raises(PlatformNotReady):
+        await async_setup_entry(hass, entry, mocked_add_entity)
```

Comment:
We want to use the core interfaces to start the tests. Set up the config entry instead.

Assert that there's no state for the entity on failure, then resolve the failure and move time forward by patching it and firing a time changed event, then assert that there's a state for the entity.